File: deepwiki_to_md/utils/error_handler.py
"""
Error Handling Utilities for DeepWiki to Markdown Converter

This module provides error handling utility functions for the DeepWiki to Markdown converter.
"""

import logging
import sys
import traceback
import requests
from typing import Optional, Callable, Any

from deepwiki_to_md.utils.localization import get_localized_message

logger = logging.getLogger(__name__)


def setup_error_handling() -> None:
    """
    Set up global error handling.
    
    This function sets up global exception handlers and other error handling mechanisms.
    """

    # Set up sys.excepthook to log uncaught exceptions
    def exception_handler(exc_type, exc_value, exc_traceback):
        if issubclass(exc_type, KeyboardInterrupt):
            # Let KeyboardInterrupt pass through
            sys.__excepthook__(exc_type, exc_value, exc_traceback)
            return

        # Log the exception
        logger.critical(
            get_localized_message("uncaught_exception"),
            exc_info=(exc_type, exc_value, exc_traceback)
        )

    sys.excepthook = exception_handler
    logger.debug("Global error handling set up")


def handle_request_error(error: requests.RequestException, url: str) -> None:
    """
    Handle a request error.
    
    Args:
        error (requests.RequestException): The request exception
        url (str): The URL that caused the error
    """
    if isinstance(error, requests.ConnectionError):
        logger.error(get_localized_message("connection_error", url=url))
    elif isinstance(error, requests.Timeout):
        logger.error(get_localized_message("timeout_error", url=url))
    elif isinstance(error, requests.HTTPError):
        status_code = error.response.status_code if error.response else "unknown"
        logger.error(get_localized_message("http_error", url=url, status_code=status_code))
    else:
        logger.error(get_localized_message("request_error", url=url, error=str(error)))


def retry_on_exception(
        func: Callable,
        max_retries: int = 3,
        retry_delay: int = 2,
        exceptions: tuple = (requests.RequestException,),
        on_retry: Optional[Callable[[Exception, int], None]] = None
) -> Callable:
    """
    Decorator to retry a function on exception.
    
    Args:
        func (Callable): Function to retry
        max_retries (int): Maximum number of retries
        retry_delay (int): Delay between retries in seconds
        exceptions (tuple): Exceptions to catch and retry on
        on_retry (Optional[Callable[[Exception, int], None]]): Function to call on retry
        
    Returns:
        Callable: Wrapped function
    """
    import time

    def wrapper(*args, **kwargs):
        last_exception = None

        for retry in range(max_retries + 1):
            try:
                return func(*args, **kwargs)
            except exceptions as e:
                last_exception = e

                if retry < max_retries:
                    if on_retry:
                        on_retry(e, retry + 1)
                    else:
                        logger.warning(
                            get_localized_message(
                                "retry_attempt",
                                function=func.__name__,
                                retry=retry + 1,
                                max_retries=max_retries,
                                error=str(e)
                            )
                        )

                    # Wait before retrying
                    time.sleep(retry_delay)
                else:
                    logger.error(
                        get_localized_message(
                            "max_retries_exceeded",
                            function=func.__name__,
                            max_retries=max_retries,
                            error=str(e)
                        )
                    )

        # If we get here, all retries failed
        raise last_exception

    return wrapper


def log_execution_time(func: Callable) -> Callable:
    """
    Decorator to log the execution time of a function.
    
    Args:
        func (Callable): Function to log execution time for
        
    Returns:
        Callable: Wrapped function
    """
    import time

    def wrapper(*args, **kwargs):
        start_time = time.time()
        result = func(*args, **kwargs)
        end_time = time.time()

        logger.debug(
            get_localized_message(
                "execution_time",
                function=func.__name__,
                time=end_time - start_time
            )
        )

        return result

    return wrapper

File: nodejs_implementation/src/utils/error-handler.js
/**
 * Error Handling Utility for DeepWiki to Markdown Converter
 * 
 * This module provides error handling functionality for the DeepWiki to Markdown converter.
 */

const { logger } = require('./logger');
const { getLocalizedMessage } = require('./localization');

/**
 * Set up global error handling.
 * 
 * This function sets up global exception handlers and other error handling mechanisms.
 */
function setupErrorHandling() {
  // Handle uncaught exceptions
  process.on('uncaughtException', (error) => {
    logger.error(getLocalizedMessage('uncaught_exception'));
    logger.error(error.stack);
    process.exit(1);
  });
  
  // Handle unhandled promise rejections
  process.on('unhandledRejection', (reason, promise) => {
    logger.error(getLocalizedMessage('unhandled_rejection'));
    logger.error(reason);
    process.exit(1);
  });
  
  logger.debug('Global error handling set up');
}

/**
 * Handle a request error.
 * 
 * @param {Error} error - The request error
 * @param {string} url - The URL that caused the error
 */
function handleRequestError(error, url) {
  if (error.code === 'ECONNREFUSED' || error.code === 'ENOTFOUND') {
    logger.error(getLocalizedMessage('connection_error', { url }));
  } else if (error.code === 'ETIMEDOUT') {
    logger.error(getLocalizedMessage('timeout_error', { url }));
  } else if (error.response) {
    const statusCode = error.response.status || 'unknown';
    logger.error(getLocalizedMessage('http_error', { url, statusCode }));
  } else {
    logger.error(getLocalizedMessage('request_error', { url, error: error.message }));
  }
}

/**
 * Retry a function on exception.
 * 
 * @param {Function} func - Function to retry
 * @param {Object} options - Retry options
 * @param {number} [options.maxRetries=3] - Maximum number of retries
 * @param {number} [options.retryDelay=2000] - Delay between retries in milliseconds
 * @param {Array<Function>} [options.retryOn=[Error]] - Errors to retry on
 * @param {Function} [options.onRetry] - Function to call on retry
 * @returns {Function} Wrapped function
 */
function retryOnException(func, options = {}) {
  const {
    maxRetries = 3,
    retryDelay = 2000,
    retryOn = [Error],
    onRetry
  } = options;
  
  return async function(...args) {
    let lastError;
    
    for (let retry = 0; retry <= maxRetries; retry++) {
      try {
        return await func(...args);
      } catch (error) {
        lastError = error;
        
        // Check if we should retry this error
        const shouldRetry = retryOn.some(errorType => error instanceof errorType);
        
        if (shouldRetry && retry < maxRetries) {
          if (onRetry) {
            onRetry(error, retry + 1);
          } else {
            logger.warn(
              getLocalizedMessage('retry_attempt', {
                function: func.name || 'anonymous',
                retry: retry + 1,
                maxRetries,
                error: error.message
              })
            );
          }
          
          // Wait before retrying
          await new Promise(resolve => setTimeout(resolve, retryDelay));
        } else if (retry === maxRetries) {
          logger.error(
            getLocalizedMessage('max_retries_exceeded', {
              function: func.name || 'anonymous',
              maxRetries,
              error: error.message
            })
          );
          throw error;
        } else {
          throw error;
        }
      }
    }
  };
}

/**
 * Measure and log the execution time of a function.
 * 
 * @param {Function} func - Function to measure
 * @returns {Function} Wrapped function
 */
function logExecutionTime(func) {
  return async function(...args) {
    const startTime = process.hrtime();
    const result = await func(...args);
    const [seconds, nanoseconds] = process.hrtime(startTime);
    const timeInMs = (seconds * 1000) + (nanoseconds / 1000000);
    
    logger.debug(
      getLocalizedMessage('execution_time', {
        function: func.name || 'anonymous',
        time: timeInMs.toFixed(2)
      })
    );
    
    return result;
  };
}

module.exports = {
  setupErrorHandling,
  handleRequestError,
  retryOnException,
  logExecutionTime
};